sound/soc/sof/mediatek/mt8195/mt8195.c

```diff
@@ -316,6 +316,19 @@ static int mt8195_dsp_suspend(struct snd_sof_dev *sdev, u32 target_state)
 {
 	struct platform_device *pdev = container_of(sdev->dev, struct platform_device, dev);
 	int ret;
+	u32 reset_sw, dbg_pc;
+
+	/* wait dsp enter idle, timeout is 1 second */
+	ret = snd_sof_dsp_read_poll_timeout(sdev, DSP_REG_BAR,
+					    DSP_RESET_SW, reset_sw,
+					    ((reset_sw & ADSP_PWAIT) == ADSP_PWAIT),
+					    SUSPEND_DSP_IDLE_POLL_INTERVAL_US,
+					    SUSPEND_DSP_IDLE_TIMEOUT_US);
+	if (ret < 0) {
+		dbg_pc = snd_sof_dsp_read(sdev, DSP_REG_BAR, DSP_PDEBUGPC);
+		dev_warn(sdev->dev, "dsp not idle, powering off anyway : swrest %#x, pc %#x, ret %d\n",
+			 reset_sw, dbg_pc, ret);
+	}
```

Comment:
I fundamentally don't understand the flow.

The decision to suspend and the call to this function is done by the kernel. I don't know how the firmware would enter idle on its own.

Your commit message shows this problem by starting with 

"
When DSP receive suspend message, DSP firmware will send interrupt to
notify host then DSP will enter idle mode
"

But there are no explanations on how the DSP receives the suspend message and it might be related to the host suspend.

I am not saying what you are doing is wrong, but only that the explanations and transitions are far from clear.

Reply:
@plbossart  Thanks for review, I have leave comment above https://github.com/thesofproject/linux/pull/3483#discussion_r826490822 to explain the suspend flow between host and DSP. not sure if it is enough or should I add them into commit message too? Thanks.

Reply:
you need to improve the commit message. A reviewer cannot possibly figure out what this patch does without a lot of additional context.

Reply:
@plbossart  thanks for your comment. I will add it.

Reply:
@plbossart  I have updated commit message. Please help to check it. Thanks.

Reply:
@yaochunhung,
Looking at the firmware code and I'm confused.
in mt8195's ipc_platform_complete_cmd() `asm volatile("waiti 15");` is done after the `clock_set_freq()` which can take long(er) time I suppose and this is what this polling code needs to wait for.
So, in SOF it is `asm volatile("waiti 15");` and in Linux `ADSP_PWAIT` is `BIT(16)` ?

I'm sure it is correct, but strange by looking at the numbers.

The commit message is still needs some tuning to get the message through:
In ipc_platform_complete_cmd() you ack/signal the reply for the host (done for all host to DSP commands), so CTX_SAVE message got tx done fine but you will continue (because ipc->pm_prepare_D3 is now set) to change a clock freq, which can take time and the host should wait for the clock frequency change. You have a signal in use for the completion of the freq change (`asm volatile("waiti 15");`) and in suspend callback on the kernel side you need to wait for it to have a clean shutdown.
Is my understanding correct?

Reply:
@ujfalusi Thanks for your suggestion. For "asm volatile("waiti 15")" is used to let DSP enter idle mode and once DSP really enter idle mode, ADSP_PWAIT(BIT 16, which is DSP internal register and our design connected it directly to register so we can check it) will be pulled high.  Basically, clock_set_freq won't take much time but I think your suggestion to move before trigger_irq_to_host_rsp is a good idea to avoid wasting time to change clock. My original idea of this patch is to handler any case which DSP is blocked by abnormal behavior or exception. Thanks.